### PR TITLE
WIP: feat(es-dev-server): allow passing chokidar options as `watch` option

### LINKED ISSUE #1398: [es-dev-server]: expose chokidar options
https://polymer.slack.com/archives/CE6D9DN05/p1583315939403900

I propose changing the type of the `watch` option from `boolean` to `boolean|WatchOptions`

https://github.com/paulmillr/chokidar/blob/6217706d11cac734c437fa4757e4b002a1d90b43/types/index.d.ts#L67-L167

### PR DESCRIPTION
**Pull request by bennypowers**
closes #1398

### CODE REVIEW

**Comment by bennypowers** [comment]
- missing tests (master skips tests)
- disconcertingly, using this branch solved my original problem (`Not Found` in browser on watch-reload) *without* passing @thepassle 's recommended options to config.

File: packages/es-dev-server/src/command-line-args.js [hunk 1]
```diff
@@ -64,7 +64,16 @@ export const commandLineOptions = [
   {
     name: 'watch',
     alias: 'w',
-    type: Boolean,
+    type: x => {
```
Comment (LarsDenBakker) [inline packages/es-dev-server/src/command-line-args.js]: I didn't know about this feature, very useful for some other cases too!

File: packages/es-dev-server/src/utils/utils.js [hunk 2]
```diff
@@ -114,6 +114,22 @@ export class SSEStream extends Stream.Transform {
   }
 }
 
+/**
+ * @param  {*}  x
+ * @return {x is boolean}
+ */
+export function isBoolean(x) {
```
Comment (LarsDenBakker) [inline packages/es-dev-server/src/utils/utils.js]: Not sure if it's worth creating a helper function for this? `typeof x === 'boolean'` is fairly straightforward?

File: packages/es-dev-server/src/utils/utils.js [hunk 3]
```diff
@@ -114,6 +114,22 @@ export class SSEStream extends Stream.Transform {
...
+  return typeof x === 'boolean';
+}
+
+/**
+ * @param  {boolean|import('chokidar').WatchOptions}  watch
+ * @return {watch is WatchOptions}
+ */
+export function isWatchOptions(watch) {
```
Comment (LarsDenBakker) [inline packages/es-dev-server/src/utils/utils.js]: This isn't used anywhere?

File: packages/es-dev-server/src/utils/utils.js (hunk 2 above)
Comment (bennypowers) [reply packages/es-dev-server/src/utils/utils.js]: except typescript makes the straightforward crooked.

the return type `x is boolean` is a necessary typeguard to let options be passed as `WatchOptions` and not `boolean|WatchOptions`

File: packages/es-dev-server/src/utils/utils.js (hunk 3 above)
Comment (bennypowers) [reply packages/es-dev-server/src/utils/utils.js]: true. see previous comment. seems typescript preferred the opposite predicate 🤷‍♂

**Comment by bennypowers** [comment]
>  create the filtewatcher with the config from the cli before we call `createServer`.

The goal is to have 
```js
// es-dev-server.config.js
module.exports = {
  // ...
  watch: { ...chokidarWatchOptions }
}
```

If we can do that the way you suggest, then sure

File: packages/es-dev-server/src/utils/utils.js (hunk 2 above)
Comment (LarsDenBakker) [reply packages/es-dev-server/src/utils/utils.js]: ok, makes sense.

**Comment by LarsDenBakker** [comment]
Initially I thought the config is cli-only, but that's not true we want this for people who use the config programmatically as well.

In that case, can we call the function parameter `watcherOrWatcherConfig` or something of that nature? And deconstruct `const fileWatcher = watcher || chokidar.watch([], ...(!isBoolean(watch) ? [watch] : []));` a little bit so that it's easier to grok at first glance :P

**Comment by bennypowers** [comment]
> Initially I thought the config is cli-only, but that's not true we want this for people who use the dev server programmatically as well.
> 
> In that case, can we call the function parameter `watcherOrWatcherConfig` 

is this what you had in mind?
```js
  let fileWatcher = watcherOrWatchOptions;
  if (!fileWatcher) {
    const chokidarOptionsArgs = !isBoolean(cfg.watch) ? [cfg.watch] : []
    fileWatcher = chokidar.watch([], ...chokidarOptionsArgs);
  }
```
in which case, `maybeWatcher` is a better name, since the config was passed in the first param.

What do we do when the config in the first param is different to the existing watcher's config? AFAIK there's no chokidar API to add or update config, so should we close the watcher and create a new one each time?

File: packages/es-dev-server/src/utils/utils.js (hunk 3 above)
Comment (bennypowers) [reply packages/es-dev-server/src/utils/utils.js]: removed

**Comment by LarsDenBakker** [comment]
Hmmm, indeed that's problematic. Introducing a breaking change for this would be unfortunate, and recreating a file watcher can create problems with people who expect their instance to be used.

Should we ignore `cfg.watch` then if it's passed as a parameter? We can just revert back to the original function parameter structure then.

**Comment by bennypowers** [comment]
@LarsDenBakker I'd appreciate some guidance on where to take this next.

**Comment by LarsDenBakker** [comment]
@bennypowers 

I think we need to go back to the original behavior of the watcher parameter where it is only a watcher, not either a watcher or options. And then we only read the options from `cfg.watch`.

**Comment by daKmoR** [comment]
are we still pursuing this? 🤔

**Comment by bennypowers** [comment]
would be nice but I don't have time to go after it atm

**Pull request closed without merging**